### PR TITLE
[JENKINS-22949] BuildTrigger & ReverseBuildTrigger should respect QueueItemAuthenticatorConfiguration

### PR DESCRIPTION
**Pull request by jglick**
# Description

See [JENKINS-22949](https://issues.jenkins-ci.org/browse/JENKINS-22949).

### Changelog entries

Proposed changelog entries:

* If you have the Authorize Project plugin installed and configured, its configuration will now be treated as final with respect to the behavior of *Job/Build* checks from **Build other projects** and **Build after other projects are built**. Formerly, if a **Per-project configurable Build Authorization** was enabled globally but some projects did not specify an **Authorization**, the two aforementioned checks would automatically fall back to checking as anonymous (typically denying build permission). To restore the former behavior, explicitly configure a **Project default Build Authorization** to be **Run as anonymous**. (Note that this will affect all build-scoped permission checks, including for example *Agent/Build*.)

### Desired reviewers

@reviewbybees & @jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/model/Executor.java [hunk 1]
```diff
@@ -398,11 +401,21 @@ public SubTask call() throws Exception {
...
+                    if (QueueItemAuthenticatorDescriptor.all().isEmpty()) {
+                        LOGGER.fine("no QueueItemAuthenticator implementations installed");
+                    } else if (QueueItemAuthenticatorConfiguration.get().getAuthenticators().isEmpty()) {
+                        LOGGER.fine("no QueueItemAuthenticator implementations configured");
+                    } else {
+                        LOGGER.log(FINE, "some QueueItemAuthenticator implementations configured but neglected to authenticate {0}", executable);
+                    }
+                }
+                try (ACLContext context = ACL.as(auth)) {
+                    queue.execute(executable, task);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Executor.java]: Queue is not overridable now, ResourceController#execute() has no permission deps. Fine

**Review by oleg-nenashev — APPROVED** [review]
LGTM, this Anonymous behavior was a serious problem when somebody was enabling the plugin. OTOH it was "as designed" 

🐝 from me, but requires feedback from @ikedam and @jenkinsci/code-reviewers since it is kinda breaking change

File: core/src/main/java/hudson/model/Executor.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Executor.java]: Right, I could not see any reason to retain the build authentication past the end of `execute`.

**Comment by ikedam** [comment]
The behavior described in [JENKINS-22949](https://issues.jenkins-ci.org/browse/JENKINS-22949) sounds better to me as it's better that the default behavior is stricter.

> It would be better if the callers of QueueItemAuthenticator.authenticate fell back to getDefaultAuthentication but replaced SYSTEM with ANONYMOUS in case there were at least one authenticator consulted. Then the replacement of SYSTEM with ANONYMOUS could be removed from both BuildTrigger and ReverseBuildTrigger. 

You can use "Run as SYSTEM" for "Project default Build Authorization" to restore the former behavior.

**Comment by jglick** [comment]
> The behavior described in JENKINS-22949

I guess you mean, the behavior I initially proposed, before JENKINS-30574 was implemented and @stephenc convinced me that this was the better approach?

> it's better that the default behavior is stricter

Meaning that you prefer the behavior prior to this PR? In most cases I would agree with the general sentiment that Jenkins should run in a strict security mode by default. In this case

* most installations are not using `QueueItemAuthenticator` at all so have nothing to lose
* there is now a clear way to configure the precise desired behavior which users would expect Jenkins to honor
* anyone affected by this change (with the exception of people who used only `ReverseBuildTrigger` and never `BuildTrigger`) would already have been seeing a stern warning printed in their build logs for years
* the former behavior was applied inconsistently: to triggers but not to other `QueueItemAuthenticator` checks, such as for `Computer.BUILD`
* the former code was hard to follow, very error-prone, and cumbersome to replicate in any other plugins which might want to apply similar checks
* I have a proposal for a monitor that would guide administrators to configure authenticators sensibly, specifically warning about an implicit fallback to `SYSTEM`

**Comment by jglick** [comment]
@reviewbybees done

continuing discussion with @ikedam

**Comment by ikedam** [comment]
@jglick 
I agree that this change makes codes clean and bahaviors consistent.
Actually I think this change is OK, but I want to know the roadmap of `QueueItemAuthenticator`:

I believe that the current inconsistency comes because:
* Unconfigured authentication is expected to have least privileges (that is, ANONYMOUS)
* The default authentication is `ACL.SYSTEM`.

The current implementaion of `BuildTrigger` is a workaround based on the former idea.

Configuring "Run as Anonymous" is to resolve the latter.
I want to know whether this is just a workaround for now and the default authentication will be ANONYMOUS in future, or it's not considered to change the default authentication and administrators that want unconfigured authentication have least privileges always need to configure "Run as Anonymous"?

**Comment by jglick** [comment]
> I want to know whether this is just a workaround for now and the default authentication will be ANONYMOUS in future

Right I think we should plan for the future to guide administrators to set a fallback to `anonymous`. See my last comment in JENKINS-24513. Possibly we should be going further and doing this in the setup wizard but there are probably some broader changes we need to make.

**Comment by ikedam** [comment]
@jglick 
Guiding in the setup wizard really sounds good.
And I believe this change doesn't cause serious security regressions as you pointed, and looks good to me.
Thanks.

File: core/src/main/java/hudson/tasks/BuildTrigger.java [hunk 2]
```diff
@@ -443,9 +410,6 @@ public FormValidation doCheck(@AncestorInPath AbstractProject project, @QueryPar
                         return FormValidation.error(Messages.BuildTrigger_NotBuildable(projectName));
                     // check whether the supposed user is expected to be able to build
                     Authentication auth = Tasks.getAuthenticationOf(project);
-                    if (auth.equals(ACL.SYSTEM) && !QueueItemAuthenticatorConfiguration.get().getAuthenticators().isEmpty()) {
-                        auth = Jenkins.ANONYMOUS; // compare behavior in execute, above
-                    }
                     if (!item.getACL().hasPermission(auth, Item.BUILD)) {
```
Comment (jglick) [inline core/src/main/java/hudson/tasks/BuildTrigger.java]: I suppose the above two lines could now be simplified to

```java
if (!item.hasPermission(Item.BUILD)) {
```

**Comment by oleg-nenashev** [comment]
Going forward with the merge according t the feedback above
